circuit_passes/src/passes/loop_unroll/loop_env_recorder.rs

```diff
@@ -0,0 +1,269 @@
+use std::cell::{RefCell, Ref};
+use std::collections::BTreeMap;
+use std::fmt::{Debug, Formatter};
+use indexmap::IndexMap;
+use compiler::intermediate_representation::BucketId;
+use compiler::intermediate_representation::ir_interface::*;
+use crate::bucket_interpreter::env::Env;
+use crate::bucket_interpreter::memory::PassMemory;
+use crate::bucket_interpreter::observer::InterpreterObserver;
+use crate::bucket_interpreter::value::Value;
+use crate::passes::GlobalPassData;
+use super::body_extractor::{UnrolledIterLvars, ToOriginalLocation};
+
+/// Holds values of index variables at array loads/stores within a loop
+pub struct VariableValues<'a> {
+    pub env_at_header: Env<'a>,
+    /// The key is the ID of the load/store bucket where the reference is located.
+    /// NOTE: uses IndexMap to preserve insertion order to stabilize test output.
+    pub loadstore_to_index: IndexMap<BucketId, (AddressType, Value)>,
+}
+
+impl<'a> VariableValues<'a> {
+    pub fn new(env_at_header: Env<'a>) -> Self {
+        VariableValues { env_at_header, loadstore_to_index: Default::default() }
+    }
+}
+
+impl Debug for VariableValues<'_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        // write!(
+        //     f,
+        //     "\n{{\n env_at_header = {}\n loadstore_to_index = {:?}\n}}",
+        //     self.env_at_header, self.loadstore_to_index
+        // )
+        write!(f, "\n  loadstore_to_index = {:?}\n", self.loadstore_to_index)
+    }
+}
+
+pub struct EnvRecorder<'a, 'd> {
+    global_data: &'d RefCell<GlobalPassData>,
+    mem: &'a PassMemory,
+    // NOTE: RefCell is needed here because the instance of this struct is borrowed by
+    //  the main interpreter while we also need to mutate these internal structures.
+    current_iter_num: RefCell<usize>,
+    safe_to_move: RefCell<bool>,
+    //NOTE: use BTreeMap instead of HashMap for consistent ordering of args in test cases
```

Comment:
The effect on the test cases is that the function arguments could come out in any order so different runs of the test would have a different argument list so that there was no predictable correlation between the order of arguments and the order of their uses within the body. For example, running twice could result in the following 2 outputs:
```
fun(a,b,c){
    use(a);
    use(b);
    use(c);
}
```
vs 
```
fun(c,a,b){
    use(a);
    use(b);
    use(c);
}
```

Reply:
Yeah, I see. Well a binary tree is it then